### PR TITLE
Protocols: Allow to compile a method with an instance of Protocol

### PR DESCRIPTION
**Pull request by jecisc**
Here is a new step for going from symbol manipulation to usage of protocols. 

With this pr we get:
- We can now give an instance of Protocol to the compiler to compile a method instead of a symbol (but the symbol works too!)
- I renamed a bunch of methods to make that explicit
- I renamed the impacted variables
- In order to know if a protocol is unclassified, we now have #isUnclassified instead of comparing to `Protocol unclassified` (Btw, this method returns a string and not a protocol, so I'll change that later probably

This will help us when we will update CompileMethod>>protocol to return a protocol. We will have no problem with the compilation and to know if this protocol is unclassified

### CODE REVIEW

File: src/OpalCompiler-Core/OpalCompiler.class.st [hunk 1]
```diff
@@ -1,30 +1,3 @@
-"
```
Comment (privat) [inline src/OpalCompiler-Core/OpalCompiler.class.st]: why did you remove the class comment?

File: src/OpalCompiler-Core/OpalCompiler.class.st [hunk 2]
```diff
@@ -558,7 +531,7 @@ OpalCompiler >> productionEnvironment: anObject [
 
 { #category : #accessing }
 OpalCompiler >> protocolName: aString [
```
Comment (privat) [inline src/OpalCompiler-Core/OpalCompiler.class.st]: update the selectors (and the parameter name).
No need to deprecate them, they are a few days old

File: src/OpalCompiler-Tests/OCTargetCompiler.class.st [hunk 3]
```diff
@@ -1,9 +1,3 @@
-"
```
Comment (privat) [inline src/OpalCompiler-Tests/OCTargetCompiler.class.st]: You do not like class comments?

File: src/System-Sources/ClassDescription.extension.st [hunk 4]
```diff
@@ -1,5 +1,16 @@
 Extension { #name : #ClassDescription }
 
+{ #category : #'*System-Sources' }
+ClassDescription >> logMethodSource: aText forMethod: aCompiledMethod inProtocol: protocol withStamp: changeStamp [
```
Comment (privat) [inline src/System-Sources/ClassDescription.extension.st:4]: do not add new methods with insane selector. Merge with the ...inProtocolNamed:.. one instead

File: src/Kernel/ClassDescription.class.st [hunk 5]
```diff
@@ -362,41 +362,41 @@ ClassDescription >> commentStamp: changeStamp [
...
+ClassDescription >> compile: sourceCode classified: protocol withStamp: changeStamp notifying: requestor logSource: logSource [
 
 	| method |
 
 	method := self compiler
 		source: sourceCode;
 		requestor: requestor;
 		failBlock:  [ ^ nil ];
-		protocolName: protocolName;
+		protocolName: protocol;
```
Comment (privat) [inline src/Kernel/ClassDescription.class.st]: you can rename the getter/setters

File: src/Kernel/Protocol.class.st [hunk 6]
```diff
@@ -38,6 +38,13 @@ Protocol >> addMethodSelector: aSymbol [
 	^ methodSelectors add: aSymbol
 ]
 
+{ #category : #converting }
+Protocol >> asSymbol [
+	"This method is here to allow compiler To compile with a Protocol"
```
Comment (privat) [inline src/Kernel/Protocol.class.st:43]: the compiler does not call `asSymbol`.
the compiler does no care. it just call the installation methods of the class with whatever was given

File: src/Kernel/String.extension.st [hunk 7]
```diff
@@ -0,0 +1,8 @@
+Extension { #name : #String }
+
+{ #category : #'*Kernel' }
+String >> isUnclassified [
+	"Return true if I am the name of Uncliassified protocol"
```
Comment (privat) [inline src/Kernel/String.extension.st]: 'Uncliassified'

Comment (privat) [reply src/Kernel/String.extension.st]: I'm not fan of this one.
Moving away from symbols and strings is good, and reifing concepts is great, but adding back random methods to string feels not nice.

File: src/OpalCompiler-Core/OpalCompiler.class.st (hunk 1 above)
Comment (jecisc) [reply src/OpalCompiler-Core/OpalCompiler.class.st]: I did not. It's probably because I loaded the change of source pointers in my image :(

I'll add it back by hand

File: src/OpalCompiler-Core/OpalCompiler.class.st (hunk 2 above)
Comment (jecisc) [reply src/OpalCompiler-Core/OpalCompiler.class.st]: Done!

File: src/Kernel/String.extension.st (hunk 7 above)
Comment (jecisc) [reply src/Kernel/String.extension.st]: I don't like it either but I'm not sure we can do without it until we transition.. Or do you have another idea?

File: src/Kernel/Protocol.class.st (hunk 6 above)
Comment (jecisc) [reply src/Kernel/Protocol.class.st:43]: Well.. It did not work before I added this method and it works now. It's probably on the methods delegated to the ClassDescription

File: src/System-Sources/ClassDescription.extension.st (hunk 4 above)
Comment (jecisc) [reply src/System-Sources/ClassDescription.extension.st:4]: I did originally but my image crashed and apparently I had a problem while loading back the code

File: src/Kernel/Protocol.class.st (hunk 6 above)
Comment (privat) [reply src/Kernel/Protocol.class.st:43]: find it and kill it! :crossed_swords:

File: src/Kernel/String.extension.st (hunk 7 above)
Comment (privat) [reply src/Kernel/String.extension.st]: A possibility: for clients that manipulate protocols either as string or Protocol (I assume only external clients, because internal one should only deal with Protocol instance?), you can transform the variable early with some `asProtocol` or something (no-op on Protocol but something on String)

Comment (jecisc) [reply src/Kernel/String.extension.st]: You prefer to have #asProtocol on String?

Comment (privat) [reply src/Kernel/String.extension.st]: yes, because `asXYZ` has a clear traditional meaning of retuning the instance of XYZ, the one of the given name. eg `asPackage` or even `asClass`.

Whereas `isUnclassified` means nothing obvious on String instances. is `'foo'` classified? what is a classified string? is there digits or Chinese characters inside?
I naturally expect that predicates (testing protocol) on strings answer relatively to intrinsic property of the string content. e.g. `isAllDigits`

Moreover, the selector `isUnclassified` gives no indication it is related to protocol in any way.

Comment (privat) [reply src/Kernel/String.extension.st]: Note: in fact, I don't know if Protocol objects are global (one Protocol instance by name on the whole system, this feels bad) or local to a class (each class has its own collection of protocols, this feels better). For the latter, `asProtocol` does not work and you should ask the targeted class to do the conversion string->protocol

File: src/Kernel/ClassDescription.class.st [hunk 8]
```diff
@@ -362,41 +362,41 @@ ClassDescription >> commentStamp: changeStamp [
...
 	nil is to be notified if an error occurs.
 	The argument code is either a string or an object that converts to a string or a PositionableStream on an object that converts to a string."
 
-	^ self compile: sourceCode classified: protocolName notifying: nil
+	^ self compile: sourceCode classified: protocol notifying: nil
 ]
 
 { #category : #compiling }
-ClassDescription >> compile: sourceCode classified: protcolName notifying: requestor [
+ClassDescription >> compile: sourceCode classified: protcol notifying: requestor [
```
Comment (Ducasse) [inline src/Kernel/ClassDescription.class.st:374]: missing o

**Comment by jecisc** [comment]
I will close this one and rework that later
